File: C#/KCube/KDC101/KDC101_Example.cs
// Title: KDC101_Example
// Created Date: 04/14/2022
// Last Modified Date: 07/18/2023
// .NET Framework version: 4.7.2
// Thorlabs DLL version: 1.14.37
// Example Description: 
// This example demonstrates how to set-up the communication for the Thorlabs 
// KDC101 controllers.


using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Threading;
using Thorlabs.MotionControl.DeviceManagerCLI;
using Thorlabs.MotionControl.GenericMotorCLI.Settings;
using Thorlabs.MotionControl.GenericMotorCLI;
using Thorlabs.MotionControl.KCube.DCServoCLI;

namespace KDC101Console
{
    class Program
    {
        static void Main(string[] args)
        {
            // Uncomment this line (and SimulationManager.Instance.UninitializeSimulations() at the end on Main)
            // If you are using a simulated device
            // SimulationManager.Instance.InitializeSimulations();

            // Enter the serial number for your device
            string serialNo = "27500125";

            DeviceManagerCLI.BuildDeviceList();


            // This creates an instance of KCubeDCServo class, passing in the Serial 
            //Number parameter.  
            KCubeDCServo device = KCubeDCServo.CreateKCubeDCServo(serialNo);

            // We tell the user that we are opening connection to the device. 
            Console.WriteLine("Opening device {0}", serialNo);

            // This connects to the device. 
            device.Connect(serialNo);

            // Wait for the device settings to initialize. We ask the device to 
            // throw an exception if this takes more than 5000ms (5s) to complete. 
            device.WaitForSettingsInitialized(5000);

            // This calls LoadMotorConfiguration on the device to initialize the 
            // DeviceUnitConverter object required for real world unit parameters.
            MotorConfiguration motorSettings = device.LoadMotorConfiguration(device.DeviceID,
            DeviceConfiguration.DeviceSettingsUseOptionType.UseFileSettings);

            // This starts polling the device at intervals of 250ms (0.25s). 

            device.StartPolling(250);

            // We are now able to Enable the device otherwise any move is ignored. 
            // You should see a physical response from your controller. 
            device.EnableDevice();
            Console.WriteLine("Device Enabled");


            // Needs a delay to give time for the device to be enabled. 
            Thread.Sleep(500);

            // Home the stage/actuator.  

            Console.WriteLine("Actuator is Homing");
            device.Home(60000);

            // Move the stage/actuator to 5mm (or degrees depending on the device 
            // connected).
            device.SetRotationModes(RotationSettings.RotationModes.RotationalRange, RotationSettings.RotationDirections.Reverse);

            decimal[] positions = {350, 355, 10};
            Console.WriteLine("Actuator is Moving");
            for (int i = 0; i<10; i++)
            {
                foreach(decimal pos in positions)
                {
                    device.MoveTo(pos, 10000);
                    Thread.Sleep(1000);
                    Console.WriteLine("Current position: %f", device.Position);
                }
            }

            //Stop polling device
            device.StopPolling();

            // Shut down controller using Disconnect() to close comms
            // Then the used library
            device.ShutDown();
            Console.WriteLine("Complete. Press any key to exit");

            // Uncomment this line if you are using Simulations
            //SimulationManager.Instance.UninitializeSimulations();

            Console.ReadKey();

        }
    }
}

File: C++/KCube/KDC101/KDC101_Example.cpp
/*
KDC101 Simple Example
Date of Creation(YYYY-MM-DD): 2021-12-24
Date of Last Modification on Github: 2022-09-21
C++ Version Used: ISO C++ 14
Kinesis Version Tested: 1.14.40
*/

#include <stdio.h>
#include <stdlib.h>
#include <conio.h>

// Include library specific header file
#include "Thorlabs.MotionControl.KCube.DCServo.h"


int __cdecl wmain(int argc, wchar_t* argv[])
{
    // Comment/uncomment this line based on whether or not you're using simulation.
	//TLI_InitializeSimulations();

	// Change this line to reflect your device's serial number
    int serialNo = 27000001; 

    // optionally set a position in device units.
    int position = 0;

	// Optionally change this value to a desired velocity (in device units/second)--*--+cx vc                                                                       
     int velocity = 0;

    // identify and access device
    char testSerialNo[16];
    sprintf_s(testSerialNo, "%d", serialNo);

    // Build list of connected device
    if (TLI_BuildDeviceList() == 0)
    {
        // get device list size 
        short n = TLI_GetDeviceListSize();
        // get KDC serial numbers
        char serialNos[100];
        TLI_GetDeviceListByTypeExt(serialNos, 100, 27);

        // Search serial numbers for given serial number
        if (strstr(serialNos, testSerialNo)) {
            printf("Device %s found", testSerialNo);
        }
        else {
            printf("Device %s Not Found", testSerialNo);
            return -1;
        }

        // open device
        if(CC_Open(testSerialNo) == 0)
        {
            // start the device polling at 200ms intervals
            CC_StartPolling(testSerialNo, 200);

			// Home device
            CC_ClearMessageQueue(testSerialNo);
            CC_Home(testSerialNo);
            printf("Device %s homing\r\n", testSerialNo);

            // wait for completion
            WORD messageType;
            WORD messageId;
            DWORD messageData;
            CC_WaitForMessage(testSerialNo, &messageType, &messageId, &messageData);
            while(messageType != 2 || messageId != 0)
            {
                CC_WaitForMessage(testSerialNo, &messageType, &messageId, &messageData);
            }

            // set velocity if desired
            if(velocity > 0)
            {
                int currentVelocity, currentAcceleration;
                CC_GetVelParams(testSerialNo, &currentAcceleration, &currentVelocity);
                CC_SetVelParams(testSerialNo, currentAcceleration, velocity);
            }

            // move to position (channel 1)
            CC_ClearMessageQueue(testSerialNo);
            CC_MoveToPosition(testSerialNo, position);
            

            // stop polling
            CC_StopPolling(testSerialNo);
            // close device
            CC_Close(testSerialNo);
        }
    }

	// Uncomment this line if you are using simulations
	//TLI_UninitializeSimulations();

    char c = _getch();
    return 0;
}
